### PR TITLE
secret/database: ensure plugins are closed if they cannot be initialized

### PR DESCRIPTION
**Pull request by briankassouf**

### CODE REVIEW

**Comment by calvn** [comment]
Shouldn't we call Database.Close() on Initialize errors for builtin database plugins as well?

File: builtin/logical/database/dbplugin/client.go [hunk 1]
```diff
@@ -18,13 +19,25 @@ type DatabasePluginClient struct {
...
 	dc.client.Kill()
 
 	return err
 }
 
+// This wraps the Initialize call and ensures we close the plugin on error.
+func (dc *DatabasePluginClient) Initialize(ctx context.Context, config map[string]interface{}, verifyConnection bool) error {
+	err := dc.Database.Initialize(ctx, config, verifyConnection)
+	if err != nil {
+		dc.Close()
```
Comment (calvn) [inline builtin/logical/database/dbplugin/client.go]: The database object in here might be nil if `verify_connection` was set to false (thus never calling `Connection()` to instantiate the object). dc.Close() -> dc.Database.Close() would be called on a nil db if the Initialize error comes from something else (e.g. https://github.com/hashicorp/vault/blob/master/plugins/helper/database/connutil/sql.go#L58).

Comment (jefferai) [reply builtin/logical/database/dbplugin/client.go]: Isn't that something that should be handled by the Close() implementation in the database though?

Comment (jefferai) [reply builtin/logical/database/dbplugin/client.go]: In fact isn't it handled already? https://github.com/hashicorp/vault/blob/master/plugins/helper/database/connutil/sql.go#L132

Comment (briankassouf) [reply builtin/logical/database/dbplugin/client.go]: Yeah the plugin implementation handles that. https://github.com/hashicorp/vault/blob/master/plugins/helper/database/connutil/sql.go#L127

Comment (calvn) [reply builtin/logical/database/dbplugin/client.go]: Ah, missed that. It's all good here!

**Pull request merged**